x/distribution/src/keeper/validator.rs

```diff
@@ -0,0 +1,149 @@
+            outstanding.rewards = outstanding.rewards.checked_sub(&rewards.rewards)?;
+            self.set_fee_pool(ctx, &fee_pool)?;
+            self.set_validator_outstanding_rewards(ctx, validator_operator_addr, &outstanding)?;
+            None
+        } else {
+            // note: necessary to truncate so we don't allow withdrawing more rewards than owed
+            Some(
+                rewards.rewards.checked_quo_dec_truncate(
+                    Decimal256::from_atomics(validator_tokens, 0)
+                        .map_err(|e| DistributionError::Numeric(e.into()))?,
+                )?,
+            )
+        };
+
+        // fetch historical rewards for last period
+        let historical = if let Some(rewards) =
+            self.validator_historical_rewards(ctx, validator_operator_addr, rewards.period - 1)?
+        {
+            rewards.cumulative_reward_ratio
+        } else {
+            return Err(DistributionError::ValidatorHistoricalRewardsNotFound(
+                validator_operator_addr.clone(),
+            ));
+        };
+
+        // decrement reference count
+        self.decrement_reference_count(ctx, validator_operator_addr, rewards.period - 1)?;
+
+        // set new historical rewards with reference count of 1
+        let cumulative_reward_ratio = if let Some(current) = current {
+            historical.checked_add(&current)?
+        } else {
+            historical
+        };
+        self.set_validator_historical_rewards(
+            ctx,
+            validator_operator_addr,
+            rewards.period,
+            &ValidatorHistoricalRewards {
+                cumulative_reward_ratio,
+                reference_count: ValidatorHistoricalRewardsReferenceCount::new(1)
+                    .expect("hardcoded value is less than upper bound"),
+            },
+        )?;
+
+        // set current rewards, incrementing period by 1
+        self.set_validator_current_rewards(
+            ctx,
+            validator_operator_addr,
+            &ValidatorCurrentRewards {
+                // TODO: empty rewards in the sdk
+                rewards: rewards.rewards,
+                period: rewards.period + 1,
+            },
+        )?;
+
+        Ok(rewards.period)
+    }
+
+    /// decrement the reference count for a historical rewards value, and delete if zero references remain
+    pub fn decrement_reference_count<DB: Database>(
+        &self,
+        ctx: &mut TxContext<'_, DB, SK>,
+        validator_operator_addr: &ValAddress,
+        period: u64,
+    ) -> Result<(), DistributionError> {
+        let mut historical = if let Some(rewards) =
+            self.validator_historical_rewards(ctx, validator_operator_addr, period)?
+        {
+            rewards
+        } else {
+            return Err(DistributionError::ValidatorHistoricalRewardsNotFound(
+                validator_operator_addr.clone(),
+            ));
+        };
+        // TODO: panics in sdk
```

Comment:
Reminder: omit usage of panic, as they do in sdk. For us, panic is programmer fault.

Reply:
I do not panic. But it may be a valid way to stop chain.

Reply:
I mean, you left comment about panic, so I mentioned it